File: core/src/me/scarlet/undertailor/engine/Collider.java
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2016 Tellerva, Marc Lawrence
 *
 * Permission is hereby granted, free of charge, to any
 * person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the
 * Software without restriction, including without
 * limitation the rights to use, copy, modify, merge,
 * publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software
 * is furnished to do so, subject to the following
 * conditions:
 *
 * The above copyright notice and this permission notice
 * shall be included in all copies or substantial portions
 * of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY
 * KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
 * THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
 * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
 * CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */

package me.scarlet.undertailor.engine;

import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.physics.box2d.Body;
import com.badlogic.gdx.physics.box2d.BodyDef.BodyType;

/**
 * Skeleton implementation for objects that can collide, and
 * thus be affected by physics.
 * 
 * <p>It is possible for {@link Collider}s to not have a
 * representative body as of yet, thus the methods are
 * required to be implemented should the body be missing
 * when called.</p>
 */
public interface Collider {

    /**
     * Returns the body of this {@link Collider}.
     * 
     * @return this Collider's body, or null if not found
     */
    Body getBody();

    /**
     * Returns the {@link BodyType} of this {@link Collider}
     * .
     * 
     * @return the BodyType of this Collider
     */
    BodyType getColliderType();

    /**
     * Sets the {@link BodyType} of this {@link Collider}.
     * 
     * @param type the new BodyType of this collider
     */
    void setColliderType(BodyType type);

    /**
     * Returns the velocity of this {@link Collider} as a
     * {@link Vector2}.
     * 
     * @return the current velocity of this Collider
     */
    Vector2 getVelocity();

    /**
     * Sets the velocity of this {@link Collider}.
     * 
     * @param xVel the new x velocity
     * @param yVel the new y velocity
     */
    void setVelocity(float xVel, float yVel);

    /**
     * Applies a force to this {@link Collider} at the
     * provided local point. 0, 0 is the center of the body.
     * 
     * <p>Forces will build up velocity when continuously
     * applied.</p>
     * 
     * <p>Applying a force anywhere other than at the body's
     * center will generate torque.</p>
     * 
     * @param x the x-coordinate of the local point to apply
     *        the force to
     * @param y the y-coordinate of the local point to apply
     *        the force to
     * @param forceX the force's power on the X axis, in
     *        newtons
     * @param forceY the force's power on the Y axis, in
     *        newtons
     */
    void applyForce(float forceX, float forceY, float x, float y);

    /**
     * Applies an impulse to this {@link Collider} at the
     * provided local point. 0, 0 is the center of the body.
     * 
     * <p>Impulses will instantly change the velocity rather
     * than build up when continuously applied.</p>
     * 
     * <p>Applying an impulse anywhere other than at the
     * body's center will generate torque.</p>
     * 
     * @param x the x-coordinate of the local point to apply
     *        the force to
     * @param y the y-coordinate of the local point to apply
     *        the force to
     * @param impX the impulse's power on the X axis, in
     *        newtons
     * @param impY the impulse's power on the Y axis, in
     *        newtons
     */
    void applyImpulse(float impX, float impY, float x, float y);

    /**
     * Applies a torque to this {@link Collider}.
     * 
     * @param torque the torque to apply
     */
    void applyTorque(float torque);

    /**
     * Notifies about a collision between this
     * {@link Collider} the provided Collider.
     * 
     * @param collider the collider to have collided with
     */
    void startCollision(Collider collider);

    /**
     * Notifies about the end of a collision between this
     * {@link Collider} and the provided Collider.
     * 
     * @param collider the collider to have ceased collision
     *        with
     */
    void endCollision(Collider collider);

    /**
     * Returns whether or not this {@link Collider} is able
     * to collide.
     * 
     * @return whether or not this Collider can collide
     */
    boolean canCollide();

    /**
     * Sets whether or not this {@link Collider} can
     * collide.
     * 
     * @param canCollide if this Collider can collide
     */
    void setCanCollide(boolean canCollide);

    /**
     * Returns the group ID of this {@link Collider}. A
     * negative group ID represents no group.
     * 
     * <p>Colliders with matching group IDs will not collide
     * with each other.</p>
     * 
     * @return the group ID of this Collider
     */
    short getGroupId();

    /**
     * Sets the group ID of this {@link Collider}. Use -1 to
     * represent no group.
     * 
     * @param id the new group ID of this collider.
     */
    void setGroupId(short id);

    // ---------------- default methods ----------------

    /**
     * Applies a force to the center of this
     * {@link Collider}.
     * 
     * @param forceX the force's power on the X axis, in
     *        newtons
     * @param forceY the force's power on the Y axis, in
     *        newtons
     */
    default void applyCenterForce(float forceX, float forceY) {
        this.applyForce(forceX, forceY, 0, 0);
    }

    /**
     * Applies an impulse to the center of this
     * {@link Collider}.
     * 
     * @param impulseX the impulse's power on the X axis, in
     *        newtons
     * @param impulseY the impulse's power on the Y axis, in
     *        newtons
     */
    default void applyCenterImpulse(float impX, float impY) {
        this.applyImpulse(impX, impY, 0, 0);
    }
}

File: source/class/collider.lua
--[[
	
	The MIT License (MIT)

	Copyright (c) 2016 Tellerva, Marc Lawrence

	Permission is hereby granted, free of charge, to any person obtaining a copy
	of this software and associated documentation files (the "Software"), to deal
	in the Software without restriction, including without limitation the rights
	to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
	copies of the Software, and to permit persons to whom the Software is
	furnished to do so, subject to the following conditions:

	The above copyright notice and this permission notice shall be included in all
	copies or substantial portions of the Software.

	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
	IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
	FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
	AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
	LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
	OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
	SOFTWARE.
	
]]--


---
-- `collider` objects are affected by physics and can interact with
-- each other in the Overworld subsystem.
--
-- @classmod collider
---

local collider = {}

---
-- Returns the body type of this `collider`.
--
-- This function can return 3 integers.
--
-- * 1 - The body is a static body.
-- * 2 - The body is a kinematic body.
-- * 3 - The body is a dynamic body.
--
-- @return the body type id of this collider
--
function collider:getColliderType() end

---
-- Sets the body type of this `collider`.
--
-- See `collider:getColliderType` for the possible type IDs.
--
-- @see collider:getColliderType
--
-- @int type the ID of the body type to set
--
function collider:setColliderType(type) end

---
-- Returns the velocity of this `collider`.
--
-- @return[1] the velocity of this collider on the x axis
-- @return[1] the velocity of this collider on the y axis
--
function collider:getVelocity() end

---
-- Sets the velocity of this `collider`.
--
-- @number velX the new velocity of this collider on the x axis
-- @number velY the new velocity of this collider on the y axis
--
function collider:setVelocity(velX, velY) end

---
-- Applies a force to this `collider`.
--
-- Applying a force at a point other than the object's point origin
-- (at local 0, 0) will generate torque, if the object's rotation is
-- not fixed (see `collider:isRotationFixed`).
--
-- @number forceX the strength of the force to apply on the x axis
-- @number forceY the strength of the force to apply on the y axis
-- @number[opt=0.0] localX the location relative to this collider's
--   origin to apply the force to on the x axis
-- @number[opt=0.0] localY the location relative to this collider's
--   origin to apply the force to on the y axis
--
function collider:applyForce(forceX, forceY, localX, localY) end

---
-- Applies an impulse to this `collider`.
--
-- Applying an impulse at a point other than the object's point
-- origin (at local 0, 0) will generate torque, if the object's
-- rotation is not fixed (see `collider:isRotationFixed`).
--
-- @number impX the strength of the impulse to apply on the x axis
-- @number impY the strength of the impulse to apply on the y axis
-- @number[opt=0.0] localX the location relative to the collider's
--   origin to apply the impulse to on the x axis
-- @number[opt=0.0] localY the location relative to the collider's
--   origin to apply the impulse to on the y axis
--
function collider:applyImpulse(impX, impY, localX, localY) end

---
-- Applies torque to this `collider`.
--
-- @number torque the strength of torque to apply
--
function collider:applyTorque(torque) end

---
-- Returns whether or not this `collider` can collide with
-- other colliders.
--
-- This only affects physical collision, where colliding objects stop
-- each other from passing in their current direction. The collider
-- will still be notified if an object collides with it, even if said
-- object passes through it due to not having physically collided.
--
-- @return whether or not the provided collider can collide
-- 
function collider:canCollide() end

---
-- Sets whether or not this `collider` can collide with other
-- colliders.
--
-- See `collider:canCollide` for the effects of this property.
--
-- @bool canCollide whether or not the provided collider should
--   collide with other objects
--
function collider:setCanCollide(canCollide) end

---
-- Returns whether or not this `collider` is affected by
-- rotational velocity, or torque.
--
-- @return whether or not the collider is affected by torque
--
function collider:isRotationFixed() end

---
-- Sets whether or not this `collider` is affected by
-- rotational velocity, or torque.
--
-- @bool rotationFixed whether or not the collider is affected by
--   torque
--
function collider:setRotationFixed(rotationFixed) end

---
-- Returns the group ID of this `collider`.
--
-- For objects that can collide, group IDs determine which objects
-- a single object may collide with. Like with the `canCollide` flag,
-- objects will still be notified if another passes through them
-- despite having collision disabled.
--
-- Group ID rules are as follows:
--
-- * Objects with the same group ID will not collide with each other.
-- * Objects with a negative group ID collide with everything.
--
-- The default group ID for any object is -1.
-- 
-- @return the group ID of the provided collider
--
function collider:getGroupId() end

---
-- Sets the group ID of this `collider`.
--
-- See `collider:getGroupId` for a brief explanation on group IDs.
--
-- @int id the new group ID of the collider
--
function collider:setGroupId(id) end
